### PR TITLE
Replace Python2 type comments with type annotations

### LINKED ISSUE #863: Replace Python 2 type comments with type annotations
Replace Python 2 type comments with type annotations for consistency with https://github.com/SeldonIO/alibi-detect/pull/725.  https://github.com/SeldonIO/alibi/pull/825 should be mergeable afterwards.

### PR DESCRIPTION
**Pull request by mauicv**
## What is this:

Removes python2 type comments with type annotations, fixes #863. Required before we can merge #825.

### CODE REVIEW

File: alibi/explainers/anchors/anchor_base.py [hunk 1]
```diff
@@ -351,7 +351,9 @@ def draw_samples(self, anchors: list, batch_size: int) -> Tuple[tuple, tuple]:
             if anchor not in self.state['t_order']:
                 self.state['t_order'][anchor] = list(anchor)
 
-        sample_stats, pos, total = [], (), ()  # type: List, Tuple, Tuple
+        sample_stats: List = []
+        pos: Tuple = tuple()
```
Comment (ascillitoe) [inline alibi/explainers/anchors/anchor_base.py:355]: Is the `tuple()` preferred to `()`?

Comment (ascillitoe) [reply alibi/explainers/anchors/anchor_base.py:355]: Just noticed we tend to mostly do `tuple()` elsewhere...

Comment (mauicv) [reply alibi/explainers/anchors/anchor_base.py:355]: I tend to use tuple because `()` is ambiguous and can create errors in certain contexts.

File: alibi/explainers/anchors/language_model_text_sampler.py [hunk 2]
```diff
@@ -60,8 +60,10 @@ def __init__(self, model: LanguageModel, perturb_opts: dict, ):
                 self.subwords_mask[vocab[token]] = True
 
         # define head, tail part of the text
-        self.head, self.tail = '', ''  # type: str, str
-        self.head_tokens, self.tail_tokens = [], []  # type: List[str], List[str]
+        self.head: str = ''
```
Comment (ascillitoe) [inline alibi/explainers/anchors/language_model_text_sampler.py:63]: Since this crops up in a few places, we probably need to decide on a convention for handling these single-line multi-definition bits with type annotations. We could do something like:

```python
self.head: str = ""; self.tail: str = ''"
```

File: alibi/explainers/cfproto.py [hunk 3]
```diff
@@ -768,7 +768,7 @@ def fit(self,
                                                  update_feature_range=False)
 
                 # combine abdm and mvdm
-                self.d_abs = {}  # type: Dict
+                self.d_abs: Any = {}
```
Comment (ascillitoe) [inline alibi/explainers/cfproto.py:771]: Why the change from `Dict` to `Any`?

File: alibi/utils/distance.py [hunk 4]
```diff
@@ -122,8 +122,8 @@ def abdm(X: np.ndarray,
     # combine dict for categorical with binned features
     cat_vars_combined = {**cat_vars, **cat_vars_bin}
 
-    d_pair = {}  # type: Dict
-    X_cat_eq = {}  # type: Dict
+    d_pair: dict = {}
```
Comment (ascillitoe) [inline alibi/utils/distance.py:125]: Just noting that elsewhere we've done `dict` -> `Dict` but here it is `Dict -> dict`

**Review by ascillitoe — APPROVED** [review]
A few comments but generally LGTM!

File: alibi/explainers/anchors/language_model_text_sampler.py (hunk 2 above)
Comment (mauicv) [reply alibi/explainers/anchors/language_model_text_sampler.py:63]: I'm happy with the above. This [PEP](https://peps.python.org/pep-0526/#global-and-local-variable-annotations) suggest placing the type hints before the multiple assignment like so:

```py
self.head: str
self.tail: str
self.head, self.tail = "", ""
```

File: alibi/explainers/cfproto.py (hunk 3 above)
Comment (mauicv) [reply alibi/explainers/cfproto.py:771]: A mistake! Good catch! Thanks

File: alibi/explainers/anchors/language_model_text_sampler.py (hunk 2 above)
Comment (ascillitoe) [reply alibi/explainers/anchors/language_model_text_sampler.py:63]: I might be being silly, but I can't actually see where they discuss multiple assignment in that PEP? Only a bit about tuple unpacking and conditional branches...

In any case, I sort of think what you posted above is more readable than what I suggested?  Although it doesn't save as many lines...

Comment (ascillitoe) [reply alibi/explainers/anchors/language_model_text_sampler.py:63]: @mauicv @jklaise what was the decision with this? Just to leave as separate lines?

Comment (jklaise) [reply alibi/explainers/anchors/language_model_text_sampler.py:63]: Didn't seem to be very important, separate lines is nice as corresponds closely to how you might define types in a dataclass.
